firestore-tos-extension/functions/src/index.ts

```diff
@@ -0,0 +1,239 @@
+      );
+    }
+
+    /** check tos documents */
+    const tosDoc = await db
+      .collection(config.collectionPath)
+      .doc("agreements")
+      .collection("tos")
+      .doc(data.tosId)
+      .withConverter(termsConverter)
+      .get();
+
+    /** Return if no agreement exists  */
+    if (!tosDoc || !tosDoc.exists) {
+      console.warn("tosDoc does not exist");
+      return;
+    }
+
+    const acknowledgementDates = {};
+
+    if (data.acknowledged) {
+      //@ts-ignore
+      acknowledgementDates.acknowledgedDate =
+        admin.firestore.FieldValue.serverTimestamp();
+    }
+
+    if (!data.acknowledged) {
+      //@ts-ignore
+      acknowledgementDates.unacknowledgedDate =
+        admin.firestore.FieldValue.serverTimestamp();
+    }
+
+    /** Set new acknowledgment */
+    const ack = {
+      ...tosDoc.data(),
+      acknowledged: data.acknowledged,
+      noticeType: data.noticeType,
+      ...acknowledgementDates,
+    };
+
+    /** Add Acknowledgment */
+    await db
+      .collection(config.collectionPath)
+      .doc("acknowledgements")
+      .collection(context.auth.uid)
+      .doc(data.tosId)
+      .withConverter(acknowledgementConverter)
+      .set(
+        { ...ack },
+        {
+          merge: true,
+        }
+      );
+
+    /** Find current acknowledgements */
+    const acknowledgements = await db
+      .collection(config.collectionPath)
+      .doc("acknowledgements")
+      .collection(context.auth.uid)
+      .withConverter(acknowledgementConverter)
+      .get()
+      .then(($) => $.docs.map((doc) => doc.data()));
+
+    /** Set claims on user and return */
+    const claims = {};
+    claims[process.env.EXT_INSTANCE_ID] = acknowledgements;
+
+    const { customClaims: existingClaims } = await auth.getUser(
+      context.auth.uid
+    );
+
+    const updatedClaims = await auth.setCustomUserClaims(context.auth.uid, {
+      ...existingClaims,
+      ...claims,
+    });
+
+    /** send event if configured */
+    await eventChannel?.publish({
+      type: "firebase.google.v1.tos-accepted",
+      data: JSON.stringify(updatedClaims),
+    });
+  }
+);
+
+export const createTerms = functions.handler.https.onCall(
+  async (data, context) => {
+    // Checking that the user is authenticated.
+    if (!context.auth) {
+      throw new functions.https.HttpsError(
+        "unauthenticated",
+        "No valid authentication token provided."
+      );
+    }
+
+    /** check if tos data has been provided  */
+    if (!data) {
+      throw new functions.https.HttpsError(
+        "invalid-argument",
+        "No valid data provided."
+      );
+    }
+
+    /** check if tos data has been provided  */
+    if (!data.noticeType) {
+      throw new functions.https.HttpsError(
+        "invalid-argument",
+        "Invalid notice type"
+      );
+    }
+
+    /** Set claims on user and return */
+    return db
+      .collection(config.collectionPath)
+      .doc("agreements")
+      .collection("tos")
```

Comment:
I think this is a bug? we shouldn't be hardcoding the tos collection?

Reply:
As we only define the top level collection path, this has been updated to `notices`. Should we have a more generic term here, or ask to define what the sub-collection would be called? Same for agreements?